### PR TITLE
RATIS-1112. Ensure a node doesn't get reelected as a leader if it vol…

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1112

### CODE REVIEW

File: ratis-server/src/test/java/org/apache/ratis/server/impl/LeaderElectionTests.java [hunk 1]
```diff
@@ -96,6 +96,27 @@ public void testChangeLeader() throws Exception {
...
+  public void testLostMajorityHeartbeats() throws Exception {
+    runWithNewCluster(3, this::runTestLostMajorityHeartbeats);
+  }
+
+  void runTestLostMajorityHeartbeats(CLUSTER cluster) throws Exception {
+    final RaftServerImpl leader = waitForLeader(cluster);
+    try {
+      isolate(cluster, leader.getId());
+      Thread.sleep(leader.getMaxTimeoutMs());
+      Thread.sleep(leader.getMaxTimeoutMs());
```
Comment (runzhiwang) [inline ratis-server/src/test/java/org/apache/ratis/server/impl/LeaderElectionTests.java:109]: `Thread.sleep(leader.getMaxTimeoutMs())` duplicated call ?

Comment (szetszwo) [reply ratis-server/src/test/java/org/apache/ratis/server/impl/LeaderElectionTests.java:109]: Yes.  The test may fail if it only sleeps once.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/FollowerState.java [hunk 2]
```diff
@@ -88,6 +93,20 @@ void stopRunning() {
     this.isRunning = false;
   }
 
+  boolean lostMajorityHeartbeats() {
```
Comment (runzhiwang) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/FollowerState.java:96]: how about rename to `isTimeoutSinceLostMajorityHeartbeats` ?

**Review by runzhiwang — APPROVED** [review]
LGTM

File: ratis-server/src/main/java/org/apache/ratis/server/impl/FollowerState.java (hunk 2 above)
Comment (szetszwo) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/FollowerState.java:96]: I got that you wanted to invert the boolean.  However, "isTimeoutSinceLostMajorityHeartbeats" is hard to understand.  So, let's keep "lostMajorityHeartbeats"?  Or I can change it to ""lostMajorityHeartbeatsRecently".

Comment (runzhiwang) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/FollowerState.java:96]: Okay, `lostMajorityHeartbeatsRecently` is better.

Comment (szetszwo) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/FollowerState.java]: Done.

**Pull request merged**